Rework Quick Start flow to avoid use of SitePicker

mpkelly opened the pull request:
Fixes https://github.com/Automattic/wp-calypso/issues/66799
Related https://href.li/?D86923-code

#### Proposed Changes

See https://github.com/Automattic/wp-calypso/issues/66799 for some background. 

Changes introduced:
1. Avoid showing the SitePicker as recommended in https://github.com/Automattic/wp-calypso/issues/66799
2. Change the wording and banner style as recommended in https://github.com/Automattic/wp-calypso/issues/66799
3. Simplify the switch statement in `<ConciergeBanner/>` - there are now just two cases: has an appointment scheduled or has sessions available (previously was three)
4. Change `<Banner/>` component style to avoid text wrapping in `<a/>` or `<button/>` elements
5. Introduce `<PurchaseListConciergeBanner/>` to make logic easier to test; added some tests
6. Show banner at `/purchases/subscriptions/:site` as well as exiting location `/me/purchases`. 

Note that the banner should only appear for `en` users until we have translations.

When the user is on SiteA and has an appointment scheduled for SiteB, they will see a banner about their upcoming appointment for SiteB. This was the existing logic which is still present. We should consider showing a banner offering to schedule an appointment for SiteA if they have one available. Please consider this during the review. 

**What the user sees when they have an appointment scheduled**:
![image](https://user-images.githubusercontent.com/6851384/187829756-c2cafe14-8ee9-432b-aafb-72e8ad58e228.png)

**What the user sees when they have sessions available**:
![image](https://user-images.githubusercontent.com/6851384/187829603-af5d4ab3-4a4b-4129-b518-e3afe47881ac.png)

#### Testing Instructions
1. Checkout this branch
2. Start your local Calypso dev environment
3. Go to the A8c Support User site, enter a username for a user with included sessions (see Phabricator Diff) and select `calypso.localhost` from the **environment** dropdown. This user has some included session
4. Start a local support session and navigate to `purchases/subscriptions/SITE_WITH_SESSION` and you should see the banner:
![image](https://user-images.githubusercontent.com/6851384/187864636-201b2296-88fa-487a-bc64-f8ceb0708680.png)
5. Click on the "schedule a date" button and verify that you go directly to the Booking Form and don't see the Site Picker, i.e. that you navigate to `me/quickstart/SITE_iD/book`
6. Next, try navigating to `/purchases/subscriptions/SITE_WITHOUT_SESSION` and confirm the banner is not shown as there are no sessions available for this site. 
8. Finally, go to `/me/purchases` and verify the banner is shown there. Also, verify that clicking `Schedule a date` takes you directly to the Booking Form (the first site ID returned in the API data will be used in this case)

You can repeat steps 3-7 with a user that has paid sessions (see Phabricator Diff).

For more info on how to find more users with Quick Start sessions, take a look at the Test Plan at D86923-code.

**I don't know how to test users that have a pending appointment - if you're reviewing this and know how to test that, then please leave a comment**

For the pending appointment tests, you're going to need to use a mock/development back end API. You will also need to set the back end up to return an upcoming appointment or an available session. For that case, verify that you see the banner while you're using `en` as your locale. Switch your locale to another language (which you can do from your account settings), and when you navigate back to the two purchases pages, verify that you don't see the Quick Start banner.

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #67186

mpkelly committed:
Rework Quick Start flow to avoid use of SitePicker

mpkelly committed:
Fix typo

pablohoneyhoney commented:
The texts can be slightly clearer, and the arrangement/layout of the JITM can be more balanced, it's a bit stuffed,  I think. cc @JanaMW27

JanaMW27 commented:
Hello,
We have a new design for these banners.
The intention is to have a more appealing and less cluttered banner with all the information but more concise.
Here is how it looks:
![quickstartsession](https://user-images.githubusercontent.com/71137829/186904538-df7fd611-e5ea-49e9-a874-e3049fb59c9f.png)

Figma file can be found [here](https://www.figma.com/file/cUSVlRgEDdxDENtWImQLVN/Quick-start-sessions-banner?node-id=3%3A37).

Let me know if you have any questions, comments, or need anything else.

daledupreez commented:
Thanks for the input, @JanaMW27!

I think we're going to need to split up getting these banners out to users and the broader update to change the shared banner design, as the latter affects many more use cases across the wp-calypso monorepo. We want to make sure we get this functional change out to users ASAP, and folding in the work to improve the banner component will push that timeline out significantly. Can you point me to any internal discussions around how we plan to push out the new banner design? I would like to make sure I am aware of those discussions before pushing forward on the work.

I also want to note some small items for the wording:
 * Our current language/branding calls these "Quick Start sessions", so we should use that language in our banners.
 * For the available sessions, some of these sessions are paid, and some are free, so it will be simpler for us to use wording like `You still have a Quick Start session available`
 * We don't have a clear end date or expiration for these sessions, which requires a much larger discussion, likely with input from our legal teams, as these sessions were originally advertised as having no end/expiration date. Perhaps we can make this a more positive CTA using language something like:
 > Schedule your Quick Start session and get one-on-one guidance from our expert Happiness Engineers to kickstart your site!

JanaMW27 commented:
Hello,
If we have to push the updated design, can we go with the new copy (plus the amendments you mentioned above)?
Thaks

mpkelly commented:
> If we have to push the updated design, can we go with the new copy (plus the amendments you mentioned above)?
> Thanks

@JanaMW27, ok - nice timing because I am just about to commit! The banner is still quite close to what you spec'd - it's just the close icon and button that would need updating later:
![image](https://user-images.githubusercontent.com/6851384/187162154-3fcf1178-ff20-49f6-be89-bb43ce6a2aea.png)

daledupreez commented:
> If we have to push the updated design, can we go with the new copy (plus the amendments you mentioned above)?

Absolutely! The idea is to get the CTA working as a first step, and then work to improve the design for the underlying component. Keeping them separate means we can get the improved copy out sooner, and possibly show the banner on the site-specific purchases page as well. And we can then tackle the bigger project of improving the shared banner component layout.

mpkelly committed:
Rework banner according to feedback; update variable name to match server-side name.

daledupreez commented:
I've just deployed the necessary back end changes in D86923-code. Hopefully that should help us get unstuck with this PR.

pablohoneyhoney commented:
Let's avoid exclamations, and also double punctuation. 

<img width="198" alt="Screen Shot 2022-08-31 at 6 07 06 PM" src="https://user-images.githubusercontent.com/25516611/187726135-6adb0c0b-a346-4167-b0a9-417e3f1a065d.png">

mpkelly committed:
Make sure Tracks event is still logged; fix punctuation.

daledupreez commented on client/components/banner/index.jsx:
```diff
@@ -114,7 +114,7 @@ export class Banner extends Component {
 	handleClick = ( e ) => {
 		const { event, feature, compact, onClick, tracksClickName, tracksClickProperties } = this.props;
 
-		if ( event && tracksClickName ) {
```
Could you share why we're removing the check for `event`? We are using it in the `cta_name` value two lines down, so this feels like it's more likely we're not using the component correctly than the component needing to skip  this specific check.

mpkelly replied:
This change should not have been committed, will reverse it.

daledupreez commented on client/me/purchases/concierge-banner/index.jsx:
```diff
@@ -2,15 +2,14 @@ import { Card } from '@automattic/components';
 	CONCIERGE_HAS_AVAILABLE_PURCHASED_SESSION,
+	CONCIERGE_HAS_UPCOMING_APPOINTMENT,
 } from 'calypso/me/concierge/constants';
-
```
Nit: I prefer keeping a new line between the component and CSS imports.

daledupreez commented on client/components/banner/style.scss:
```diff
@@ -217,6 +217,11 @@
 		}
 	}
 
+	a,
+	button {
+		white-space: nowrap;
+	}
```
I am really worried that this may introduce unexpected behaviour elsewhere in Calypso, so I don't think we should add these styles to the banner component in this PR.

If this style is really needed, then I think we should add a `className` to the banner and add the style specifically for our context.

mpkelly replied:
yes, good point! I think we need it to avoid weird text wrapping on the button, but I will make it specific to this ConciergeBanner.

daledupreez commented on client/me/purchases/concierge-banner/index.jsx:
```diff
@@ -42,86 +42,91 @@ class ConciergeBanner extends Component {
-		let mainText;
+		let title;
+		let description;
 		let buttonText;
```
Thought for our next iteration: it would be good to work out if we can add some basic icons to the banners to add a bit of colour/pop and context.

daledupreez commented on client/me/purchases/concierge-banner/index.jsx:
```diff
@@ -42,86 +42,91 @@ class ConciergeBanner extends Component {
+					callToAction={ buttonText }
+					description={ description }
+					dismissPreferenceName={ `quick-start-banner-${ bannerType }` }
+					href={ `/me/quickstart/${ quickStartSite }` }
```
Might it be cleaner to explicitly adjust the path based on whether `quickStartSite` is defined? e.g. 
```
quickStartUrl = '/me/quickstart' . ( quickStartSite ? `/${ quickStartSite }` : '' );
```
My concern here is that I am not sure the URL with a trailing slash will be redirected/handled correctly.

daledupreez replied:
Follow-up: I think this should generally be the site slug instead of the siteId. So we should use the `getSiteSlug()` selector based on the value of `siteId` to build the URL. The site ID works, but we almost always use the site slug in our URLs.

mpkelly replied:
We should always have a `quickStartSite` when this component gets rendered now, because of the `shouldShowBanner` check in `<PurchaseListConciergeBanner/>`. There is a good case to merge these two components into one, as it would make the logic clearer. Maybe that could be done in the next iteration. 

I have made it use the slug for now.

mpkelly replied:
The code is now using `getSiteSlug`. There is no need for conditional logic here as we will always have a site if the Banner is shown.

daledupreez commented on client/me/purchases/concierge-banner/index.jsx:
```diff
@@ -42,86 +42,91 @@ class ConciergeBanner extends Component {
-			this.getBannerContent();
+		const { buttonText, description, title, event } = this.getBannerContent();
+
+		const referrer = location.pathname.startsWith( '/me/purchases' )
+			? '/me/purchases'
+			: '/purchases/subscriptions/:site';
```
This doesn't feel right to me. I think we should not mediate the current path directly, and we should probably not be using `location` directly either.

I _think_ it will probably be sufficient for us to handle this via the `getCurrentRoute()` and `getCurrentRouteParameterized()` selectors (where we'd use the second when we `sited` is defined).

mpkelly replied:
I've changed this as suggested.

daledupreez commented on client/me/purchases/purchases-list/purchase-list-concierge-banner.tsx:
```diff
@@ -0,0 +1,52 @@
+import React from 'react';
```
Do we need this import? I think we handle injecting React via our build tools.)

daledupreez commented on client/me/purchases/concierge-banner/index.jsx:
```diff
@@ -42,86 +42,91 @@ class ConciergeBanner extends Component {
+				event = 'view-concierge-dashboard';
 				break;
-
+			case CONCIERGE_HAS_AVAILABLE_PURCHASED_SESSION:
 			case CONCIERGE_HAS_AVAILABLE_INCLUDED_SESSION:
```
Might it be worth creating a dedicated constant to represent `CONCIERGE_HAS_AVAILABLE_SESSION`? We may or may not be able to collapse the usages, but it would be simpler if we aren't distinguishing between these two cases any more.

daledupreez commented on client/me/purchases/purchases-list/purchase-list-concierge-banner.tsx:
```diff
@@ -0,0 +1,52 @@
+	if ( nextAppointment ) {
+		bannerType = CONCIERGE_HAS_UPCOMING_APPOINTMENT;
+	} else {
+		bannerType = CONCIERGE_HAS_AVAILABLE_PURCHASED_SESSION;
```
Relating to my comment RE collapsing the banner type constants, this usage feels weird because we don't know if the session was purchased or included.

mpkelly replied:
I will refactor these constants too. I think `CONCIERGE_HAS_UPCOMING_APPOINTMENT` and `CONCIERGE_HAS_AVAILABLE_SESSION` should be ok - the distinction between purchased or included isn't really important on the front-end.

daledupreez commented on client/me/purchases/purchases-list/test/purchase-list-concierge-banner.tsx:
```diff
@@ -0,0 +1,94 @@
+import { PurchaseListConciergeBanner } from 'calypso/me/purchases/purchases-list/purchase-list-concierge-banner';
+import { createReduxStore } from 'calypso/state/index';
+
+describe( 'PurchaseListConciergeBanner tests', () => {
```
💖 for the tests!

daledupreez commented on client/my-sites/purchases/controller.js:
```diff
@@ -23,7 +24,12 @@ export function redirectToPurchases( context ) {
-	context.primary = <Purchases />;
+	context.primary = (
+		<>
+			<QueryConciergeInitial />
```
I think it's better to pull the Query* component into the `Purchases` component itself, as it's much easier to see how its being used in context. Pushing it up into the controller feels like it makes the wiring even more magical, and may make it more complicated to work out how the `Purchases` component lifecycle will interact with the query component.

mpkelly replied:
I think I tried it there originally and it didn't work, hence putting it here. I will retry though.

mpkelly replied:
Well, it's working after I moved down into the `Purchases` component, so will commit a change.

daledupreez commented on client/my-sites/purchases/main.tsx:
```diff
@@ -50,6 +54,10 @@ export function Purchases() {
 	const translate = useTranslate();
 	const siteSlug = useSelector( getSelectedSiteSlug );
 	const logPurchasesError = useLogPurchasesError( 'site level purchases load error' );
+	const nextAppointment = useSelector( getConciergeNextAppointment );
+	const isUserBlocked = useSelector( getConciergeUserBlocked );
+	const availableSessions = useSelector( getAvailableConciergeSessions );
```
Naming nits: it's not clear from the ~variable~ const names what the items refer to. Given that we're on a payment-related page, `isUserBlocked` is almost certainly not specific enough.

I'd vote for more specific names for all three, but especially for `isUserBlocked`.
```suggestion
	const nextConciergeAppointment = useSelector( getConciergeNextAppointment );
	const isConciergeUserBlocked = useSelector( getConciergeUserBlocked );
	const availableConciergeSessions = useSelector( getAvailableConciergeSessions );
```

daledupreez commented on client/state/selectors/get-available-concierge-sessions.js:
```diff
@@ -0,0 +1,7 @@
+import { get } from 'lodash';
+
+import 'calypso/state/concierge/init';
+
+export default ( state ) => {
+	return get( state, 'concierge.availableSessions', [] );
+};
```
We're trying to remove `lodash` from the codebase, so we should use native Javascript for this. Also, we could use TypeScript for this new file.
```suggestion
import 'calypso/state/concierge/init';

export default ( state ) => {
	return state?.concierge?.availableSessions ?? [];
};
```

daledupreez reviewed (commented):
I have some notes from inspection up. Apologies for leaving that so long!

daledupreez commented on client/me/purchases/concierge-banner/index.jsx:
```diff
@@ -42,86 +42,91 @@ class ConciergeBanner extends Component {
-				mainText = translate(
-					'Get ready with your questions for your upcoming Quick Start session appointment.',
+				title = translate( 'Your session is coming up soon' );
+				description = translate(
+					'Your {{supportLink}}Quick Start support session{{/supportLink}} is approaching. Get ready for your one-to-one with our Happiness Engineer.',
```
Another opportunity for later iteration: we can show the date and time for the next scheduled session, and possibly the site the session is for.

daledupreez reviewed (commented):
I've now tested and the functional part looks good for the various cases. As per our side conversation, I noticed that we don't have any users with upcoming appointments at the moment, so we needed to provide mock appointment data from the API.

mpkelly committed:
Address review comments

mpkelly commented:
@daledupreez, I addressed those review comments in my last commit. 

On the subject of the next iteration, I think the purchases should have a "Schedule a date" button or link somewhere too. Here's what you see at `/me/purchases` if you paid for sessions (two in this case):

**List**
![image](https://user-images.githubusercontent.com/6851384/188053380-54d9783b-a123-443f-b092-8941e5f059c9.png)

**Detail**
![image](https://user-images.githubusercontent.com/6851384/188053639-cce062f8-0852-4b0b-acad-5101e23082b5.png)

I think there should be a way to use these purchases - at least on the Detail page.

daledupreez commented on client/me/purchases/concierge-banner/index.jsx:
```diff
@@ -2,24 +2,27 @@ import { Card } from '@automattic/components';
+			CONCIERGE_HAS_AVAILABLE_SESSION,
 		] ).isRequired,
-		sites: PropTypes.array.isRequired,
+		availableSessions: PropTypes.array.isRequired,
```
Nit: it would also be good to document the `showPlaceholder`, `nextAppointmentSiteId`, and `siteId` props too.

mpkelly replied:
I've done that one too.

daledupreez commented on client/me/purchases/concierge-banner/index.jsx:
```diff
@@ -119,7 +112,7 @@ class ConciergeBanner extends Component {
 					description={ description }
 					dismissPreferenceName={ `quick-start-banner-${ bannerType }` }
-					href={ `/me/quickstart/${ quickStartSite }` }
+					href={ `/me/quickstart/${ quickStartSiteSlug }` }
```
To support more specific CSS, please specify a CSS class. (Location is only because it was close to an existing edit.)
```suggestion
					href={ `/me/quickstart/${ quickStartSiteSlug }` }
					className="concierge-banner"
```

mpkelly replied:
I added that class and a more specific rule.

daledupreez commented on client/me/purchases/concierge-banner/style.scss:
```diff
@@ -1,51 +1,58 @@
+	}
+}
+
+.banner__action {
```
Please wrap this selector, otherwise it can lead to hard-to-diagnose Heisenbugs that depend on the load sequence for different components. We'll need to use the `className` prop on the `Banner` component to support that.
```suggestion
.concierge-banner .banner__action {
```

daledupreez replied:
I am also not sure why we need this rule. Could you share the reason for it? My understanding is that we should let long button text wrap and result in multi-line buttons when that is necessary.

daledupreez replied:
Also, if we do need to keep this rule, then I think we need to apply it only for larger screen sizes, as we should let the text overflow on narrower screens/mobile, where we default to having the CTA span the entire banner below the main banner text.

mpkelly replied:
We discussed this in Slack. To avoid button text wrapping, we will use `min-width: 120px` as this text could be quite long in some languages and `whitespace: nowrap` might not always be appropriate.

daledupreez commented on client/me/purchases/purchases-list/test/purchase-list-concierge-banner.tsx:
```diff
@@ -76,16 +78,23 @@ describe( 'PurchaseListConciergeBanner tests', () => {
-		expect( screen.queryByText( 'Your session is coming up soon' ) ).toBeInTheDocument();
+		expect(
+			screen.queryByText( 'Your Quick Start session is coming up soon' )
+		).toBeInTheDocument();
```
Question: do we want to verify that the right link is being shown as well?

mpkelly replied:
I've added the URL checks

daledupreez commented on client/state/selectors/get-available-concierge-sessions.ts:
```diff
@@ -0,0 +1,4 @@
+import 'calypso/state/concierge/init';
+import type { AppState } from 'calypso/types';
+
+export default ( state: AppState ) => state?.concierge?.availableSessions || [];
```
Nice and clean! I like! 😍

daledupreez commented on client/components/banner/style.scss:
```diff
@@ -261,7 +261,8 @@
 	margin-top: -8px;
 }
 
-.is-horizontal .dismissible-card__close-button, .is-horizontal .banner__close-icon {
```
Minor nit: it may be worth reverting the CSS change to this file, as it doesn't feel right to include only a linting update in this otherwise unrelated PR.

mpkelly replied:
I've done that.

daledupreez commented:
> I think there should be a way to use these purchases - at least on the Detail page.

I fully agree! I know we have links from the detail page to management UIs for domains and email products, so we could easily add a similar link for Quick Start sessions. They aren't super-visible CTAs, but we can at least surface something actionable.

daledupreez reviewed (changes requested):
I have a few more comments from inspection, will be testing shortly.

The only blocking issue is the CSS for the CTA in the banner, which should be pretty easy to address.

daledupreez commented:
Testing-wise, things look good. So we should be able to get this shipped soon!

mpkelly committed:
Address review comments:

- add url check to tests
- revert all CSS changes from banner/style.css
- use min-width instead of whitespace

mpkelly committed:
Document external props

daledupreez commented on client/me/purchases/concierge-banner/index.jsx:
```diff
@@ -107,6 +110,7 @@ class ConciergeBanner extends Component {
 			<>
 				<TrackComponentView eventName="calypso_purchases_concierge_banner_view" />
 				<Banner
+					className={ 'concierge-banner' }
```
Style nit: this isn't a dynamic value, so you can just use `"concierge-banner"` as a string.
```suggestion
					className="concierge-banner"
```

mpkelly replied:
This would be a good candidate for a `prettier` rule. Currently, it just maintains what exists already, which can be with or without brackets. I've manually fixed it and a couple of others for now.

daledupreez committed:
Hide banner if we don't have translations

daledupreez commented:
Overall, things are looking really good from my testing. I did add a simple check for missing translations, so we can roll this out immediately, and will start showing the banner in other languages when the translations are ready. The change is in https://github.com/Automattic/wp-calypso/pull/67024/commits/e96a87a628f0cfa06ee689dbe2c564a598dfbf5f - I'd love you to smoke test that aspect of this PR when you get in on Monday, @mpkelly.

daledupreez reviewed (approved):
I am going to approve this with the assumption that @mpkelly will test my translation-related change before shipping this (which will likely be on Monday).

mpkelly committed:
Update locale check to handle country-specific locales; change formatting of string props

mpkelly commented:
@daledupreez, I had to make a small change to your translation logic as my locale is being reported as `en-gb` which means the check failed and I saw nothing.

mpkelly commented:
This has been released. 🚀

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7511588

Thank you @mpkelly for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7511588).
